src/databricks/sql/thrift_backend.py

```diff
@@ -260,19 +265,40 @@ def make_request(self, method, request):
         def get_elapsed():
             return time.time() - t0
 
+        def make_bounded_delay(attempt: int, delay: int):
+            """Bound delay (seconds) by [min_delay*1.5^(attempt-1), max_delay]"""
+            delay = max(delay, self._retry_delay_min * math.pow(1.5, attempt - 1))
+            delay = min(delay, self._retry_delay_max)
+            return delay
+
         def extract_retry_delay(attempt):
-            # encapsulate retry checks, returns None || delay-in-secs
-            # Retry IFF 429/503 code + Retry-After header set
+            """Returns Retry-After header contents IFF 429/503 code + Retry-After header set"""
```

Comment:
As we have learned that the THttpClient doesn't seem reset the fields for code and headers before each request, we can read values from a previous request in case an exception occurs before `flush` is reached. See: https://github.com/apache/thrift/blob/7a5f29d48d6b99f14190d9d7bff0e63170965087/lib/py/src/transport/THttpClient.py#L191-L194
I think your fix will still work fine, but we might want to reset these explicitly to prevent potential confusion later.

Reply:
Thank you. I'll make a follow-up PR that does this.